herd/AArch64Sem.ml

```diff
@@ -1055,30 +1028,100 @@ module Make
+
+      let lift_fault_memtag mfault mm dir ii =
+        if has_handler ii then
+          fun ma -> M.bind_ctrldata ma (fun _ -> mfault >>! B.Fault dir)
+        else
+          let open Precision in
+          match C.precision,dir with
+          | (Fatal,_)|(LoadsFatal,(Dir.R)) ->
+             fun ma ->  ma >>*= (fun _ -> mfault >>! B.Exit)
+          | (Handled,_)|(LoadsFatal,Dir.W) ->
+             fun ma ->
+             let (>>) = M.bind_ctrl_first_outputs in
+             let ma = ma >> (fun a -> mfault >>! a) in
+             mm ma >>! B.Next []
+          | Skip,_ ->
+             fun _ ->
+               Warn.fatal "Memtag extension has no 'Skip' fault handling mode"
+
+      let lift_memtag_phy rA dir updatedb mop ma an ii =
+        let lift_tag =
+          let mphy = fun ma a_virt ->
+            M.delay_kont "4" ma
+              (fun a_phy ma ->
+                 delayed_check_tags a_virt (Some a_phy) ma ii
+                   (fun ma -> ma >>= M.ignore >>= B.next1T)
+                   (fun ma -> ma >>! B.Fault dir)) in
+          lift_kvm Dir.R false true mop ma an ii mphy in
+        let phy_addr_reg = AArch64.Internal 0 in
+        let lift_data =
+          let mphy = fun ma _ -> ma >>= fun a -> M.unitT (B.Next [(phy_addr_reg, a)]) in
+          lift_kvm dir updatedb false mop ma an ii mphy in
+        let mphy mtag mdata =
+          let get_phy_addr a_virt = function
+            | B.Next bds -> begin
+                try M.unitT (List.assoc phy_addr_reg bds)
+                with Not_found -> M.unitT a_virt
+              end
+            | _ -> Warn.fatal "Cannot handle argument to get_phy_addr" in
+          let ma = ma >>= fun a_virt -> M.aslseq mtag (fun _ -> mdata) >>=
+            get_phy_addr a_virt in
+          M.delay_kont "lift_memtag_phy addr" ma
+            (fun a ma ->
+               let ac =
+                 match Act.access_of_location_std (A.Location_global a) with
+                 | Access.PTE as ac -> ac
+                 | _ -> Access.PHY in
+               mop ac ma >>= M.ignore >>= B.next1T) in
+        let mfault mtag mdata =
+          let mm = mop Access.PHY in
+          let ft = Some FaultType.AArch64.TagCheck in
+          let fault a_virt = lift_fault_memtag
+              (mk_fault (Some a_virt) dir an ii ft None) mm dir ii in
+          ma >>= fun a_virt -> (mtag >>| mdata) >>*= fun _ -> fault a_virt mzero in
+        let is_next = function | B.Next _ -> V.one | _ -> V.zero in
+        let m = M.delay_kont "lift_memop data" lift_data
+            (fun data_op mdata_op ->
+               M.choiceT (is_next data_op)
+                 (M.delay_kont "lift_memop tag" lift_tag
+                    (fun tag_op mtag_op ->
+                       M.choiceT (is_next tag_op)
+                         (mphy mtag_op mdata_op)
+                         (mfault mtag_op mdata_op)))
+                 (ma >>= fun _ -> mdata_op)) in
+        M.short3 (is_this_reg rA) (E.is_commit) m
```

Comment:
This call was adding an iico_data edge from the register read of rA (address register) to the "color" branching event. It was quite brittle because it was relying on having three  steps of iico_data between those two events. It is no longer working because there are now two events betwen those two events.
![A001](https://github.com/herd/herdtools7/assets/5621641/ba022919-aa0b-4bc7-8448-5326cb60c0c9)
An example of this added edge can be seen on the diagram above from event h to event j. Notice that the edge  means that the comparison is between the tag from the virtual address read from `rA` and `tag(x)`.

Reply:
Ah I missed that, I understood the point of M.short3, but I didn't realise that it doesn't work. I think we could solve this by generalizing M.short3 to be M.short (add an iico_data regardless of the number of events in between), and changing E.is_commit to be more specific to make sure we're catching the event that checks for tag match. Does that sound reasonable?

Reply:
Actually it seems to work, or maybe I didn't understand that the problem was. Locally, for:
```
AArch64 STR-TaggedNormal
{
 [PTE(x)]=(oa:PA(x),attrs:(TaggedNormal));
 0:X0=x:green;
}
 P0          ;
 MOV W1,#1   ;
L0:          ;
 STR W1,[X0] ;
exists(x=0)
```

I get:

![image](https://github.com/herd/herdtools7/assets/224592/9b2c8f4f-48ff-4c1d-9d9d-73af18860f8f)

Which has an `iico_data` between `l` and `k`.

Reply:
Ah I may be wrong I have to double-check. I like your idea of a more specific `E.is_commit`. This would result in a less brittle technique for adding this `iico_data` edge. One could  match on the message, or add some field to branching events.

Reply:
I have checked, sorry for teh wrong alert.

Reply:
I've implemented what we discussed above in two steps:

- Improve `E.is_commit` and make it more specific
- Add a new monad `M.short p1 p2 s` that adds `iico_data` starting from any event that satisfies `p1` and finish on any event that satisfies `p2`

However these two changes strictly speaking are not necessary so they are for us to decide if they are less brittle or not. Also now `short3` is not used anywhere.